src/Cardano/Wallet/Binary.hs

```diff
@@ -428,11 +428,19 @@ decodeTxWitness = do
     _ <- CBOR.decodeListLenCanonicalOf 2
     t <- CBOR.decodeWord8
     case t of
-        0 -> CBOR.decodeTag *> (PublicKeyWitness <$> CBOR.decodeBytes)
-        1 -> CBOR.decodeTag *> (ScriptWitness <$> CBOR.decodeBytes)
-        2 -> CBOR.decodeTag *> (RedeemWitness <$> CBOR.decodeBytes)
-        _ -> fail
+        0 -> do
+            _ <- CBOR.decodeTag
+            bs <- CBOR.decodeBytes
+            case CBOR.deserialiseFromBytes decodePKWitness (BL.fromStrict bs) of
+                Left err -> fail $ show err
+                Right (_, input) -> return input
+        _ -> error
             $ "decodeTxWitness: unknown tx witness constructor: " <> show t
```

Comment:
This looks wrong for two reasons:

- We should really use `fail` here in case of problems (`CBOR.Decoder` is a `MonadFail`) instead of a wild runtime exception.
- There are some `RedeemWitness` (coming from redemption certificats) on the chain so we can't fail when parsing such transactions.

Reply:
Yes.

But, with `fail` here I get infinite / so large output I have to close my entire shell 🤔 

```
Cardano.WalletSpec
  A newly created wallet can sync with the chain FAILED [1]
Cardano.Wallet.Network.HttpBridge
  Happy paths
    get from packed epochs FAILED [2]
    get from packet epochs and filter by start slot FAILED [3]

Failures:

  test/integration/Cardano/WalletSpec.hs:53:13:
  1) Cardano.WalletSpec A newly created wallet can sync with the chain
       predicate failed on: SlotId {epochNumber = 0, slotNumber = 0}

  To rerun use: --match "/Cardano.WalletSpec/A newly created wallet can sync with the chain/"

  test/integration/Cardano/Wallet/Network/HttpBridgeSpec.hs:58:9:
  2) Cardano.Wallet.Network.HttpBridge, Happy paths, get from packed epochs
       uncaught exception: ServantError
       DecodeFailure "DeserialiseFailure 1013 \"expected bytes\"" (Response {responseStatusCode = Status {statusCode = 200, statusMessage = "OK"}, responseHeaders = fromList [("Content-Length","20651648"),("Content-Type","text/plain"),("Date","Sun, 28 Apr 2019 22:24:27 GMT")], responseHttpVersion = HTTP/1.1, responseBody = "\254CARDANOPACK\NUL\NUL\NUL\SOH\NUL\t\227\155\130\NUL\131\133\SUB-\150J\tX \221_-@n\166\198+n7\SI\137\204\&9\216\SO@\NAK-[\NAK7\196\DC3\135\&5\158\"\r\223_\146X C>h\163\&8\ETB\133\253\195h(H#~\218\146V\197\198\133\&2\175\ETB\251\226\212\190\&5\191\247B\169\130\SO\129\SUB\NUL\EOT\157\NAK\129\160\159X\FSe\144J\137\230\208\229\248\129Q=\ETB6\148^\ENQ\ESCv\240\149\236\161\&8\238\134\157T=X\FS\175(\NUL\193$\229\153\214\222\193\136\167_\139\253\227\151\235\183x\SYN:\CAN$\ETXq\242\209X\FS\GS\235\130\144\132\STX\199\238>\254\177o6\157\151\251\163\SYN\238b\GS\t\179+\137i\229KX\FS\175(\NUL\193$\229\153\214\222\193\136\167_\139\253\227\151\235\183x\SYN:\CAN$\ETXq\242\209X\FS\175(\NUL\193$\229\153\214\222\193\136\167_\139\253\227\151\235\183x\SYN:\CAN$\ETXq\242\209X\FST\DC1\199\191\135\194R`\152\&1\163\&7\167\DC3\228\133\150h\203\167\187\167\n\156>\247\195\152X\FSPq\216\128-\221\ENQ\197\159M\185\a\189\ETBI\232.bB\202\246Q+ \168\&6\143\207X\FS\GS\235\130\144\132\STX\199\238>\254\177o6\157\151\251\163\SYN\238b\GS\t\179+\137i\229KX\FST\DC1\199\191\135\194R`\152\&1\163\&7\167\DC3\228\133\150h\203\167\187\167\n\156>\247\195\152X\FSPq\216\128-\221\ENQ\197\159M\185\a\189\ETBI\232.bB\202\246Q+ \168\&6\143\207X\FS\175(\NUL\193$\229\153\214\222\193\136\167_\139\253\227\151\235\183x\SYN:\CAN$\ETXq\242\209X\FST\DC1\199\191\135\194R`\152\&1\163\&7\167\DC3\228\133\150h\203\167\187\167\n\156>\247\195\152X\FSe\144J\137\230\208\229\248\129Q=\ETB6\148^\ENQ\ESCv\240\149\236\161\&8\238\134\157T=X\FS\GS\235\130\144\132\STX\199\238>\254\177o6\157\151\251\163\SYN\238b\GS\t\179+\137i\229KX\FSPq\216\128-\221\ENQ\197\159M\185\a\189\ETBI\232.bB\202\246Q+ \168\&6\143\207X\FST\DC1\199\191\135\194R`\152\&1\163\&7\167\DC3\228\133\150h\203\167\187\167\n\156>\247\195\152X\FSPq\216\128-\221\ENQ\197\159M\185\a\189\ETBI\232.bB\202\246Q+ \168\&6\143\207X\FSe\144J\137\230\208\229\248\129Q=\ETB6\148^\ENQ\ESCv\240\149\236\161\&8\238\134
```

Reply:
That is weird :thinking: ...